src/native/libs/System.Security.Cryptography.Native.Apple/pal_swiftbindings.swift

```diff
@@ -4,166 +4,168 @@
-    aadLength: Int32
-) -> Int32 {
-    let nonceData = Data(bytesNoCopy: noncePtr, count: Int(nonceLength), deallocator: Data.Deallocator.none)
-    let key = Data(bytesNoCopy: keyPtr, count: Int(keyLength), deallocator: Data.Deallocator.none)
-    let ciphertext = Data(bytesNoCopy: ciphertextPtr, count: Int(ciphertextLength), deallocator: Data.Deallocator.none)
-    let aad = Data(bytesNoCopy: aadPtr, count: Int(aadLength), deallocator: Data.Deallocator.none)
-    let tag = Data(bytesNoCopy: tagPtr, count: Int(tagLength), deallocator: Data.Deallocator.none)
-    let symmetricKey = SymmetricKey(data: key)
+    static func seal<Plaintext>(_ plaintext: Plaintext, using key: SymmetricKey, nonce: SealedBox.Nonce?) throws -> SealedBox where Plaintext: DataProtocol
+    static func seal<Plaintext, AuthenticatedData>(_ plaintext: Plaintext, using key: SymmetricKey, nonce: SealedBox.Nonce?, authenticating additionalData: AuthenticatedData) throws -> SealedBox where Plaintext: DataProtocol, AuthenticatedData: DataProtocol
+    static func open<AuthenticatedData>(_ sealedBox: SealedBox, using key: SymmetricKey, authenticating additionalData: AuthenticatedData) throws -> Data where AuthenticatedData: DataProtocol
+    static func open(_ sealedBox: SealedBox, using key: SymmetricKey) throws -> Data
+}
 
-    guard let nonce = try? ChaChaPoly.Nonce(data: nonceData) else {
-        return 0
-    }
+extension AES.GCM.Nonce: NonceProtocol {}
+extension AES.GCM.SealedBox: SealedBoxProtocol {
+    typealias Nonce = AES.GCM.Nonce
+}
+extension AES.GCM: AEADSymmetricAlgorithm {}
 
-    guard let sealedBox = try? ChaChaPoly.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag) else {
-        return 0
-    }
+extension ChaChaPoly.Nonce: NonceProtocol {}
+extension ChaChaPoly.SealedBox: SealedBoxProtocol {
+    typealias Nonce = ChaChaPoly.Nonce
+}
+extension ChaChaPoly: AEADSymmetricAlgorithm {}
 
-    do {
-        let result = try ChaChaPoly.open(sealedBox, using: symmetricKey, authenticating: aad)
+func encrypt<Algorithm>(
+    _ algorithm: Algorithm.Type,
+    key: UnsafeBufferPointer<UInt8>,
+    nonceData: UnsafeBufferPointer<UInt8>,
+    plaintext: UnsafeBufferPointer<UInt8>,
+    cipherText: UnsafeMutableBufferPointer<UInt8>,
+    tag: UnsafeMutableBufferPointer<UInt8>,
+    aad: UnsafeBufferPointer<UInt8>) throws where Algorithm: AEADSymmetricAlgorithm {
 
-        assert(plaintextBufferLength >= result.count)
-        result.copyBytes(to: plaintextBuffer, count: result.count)
-        return 1
-    }
-    catch CryptoKitError.authenticationFailure {
-        return -1
-    }
-    catch {
-        return 0
-    }
+    let symmetricKey = SymmetricKey(data: key)
+
+    let nonce = try Algorithm.SealedBox.Nonce(data: nonceData)
+
+    let result = try Algorithm.seal(plaintext, using: symmetricKey, nonce: nonce, authenticating: aad)
+
+    _ = result.ciphertext.copyBytes(to: cipherText.baseAddress!, count: cipherText.count)
```

Comment:
This seems a tad concerning that we have created instances of types that we cannot actually use as intended.

Should we consider using a custom swift struct with a pointer and a length as members? That way we don't have a struct instance that, when actually used as intended, cause problems.

Alternatively I wonder if `UnsafeMutableRawBufferPointer` is a better choice (which is what the bindings were using before)

Reply:
So I think that there's some bugs in the Swift implementation of these methods, as it should gracefully handle all of these cases as I had them previously. I'll keep investigating to see what we can do to actually use these types as expected.

Reply:
I've narrowed the issues I was seeing to a bug in Swift, which I've filed. I've updated this PR with a workaround